streams/src/main/java/org/apache/kafka/streams/kstream/internals/MaterializedInternal.java

```diff
@@ -29,14 +31,26 @@
     private final boolean queryable;
 
     public MaterializedInternal(final Materialized<K, V, S> materialized) {
-        this(materialized, null, null);
+        this(materialized, null, null, null);
     }
 
     public MaterializedInternal(final Materialized<K, V, S> materialized,
                                 final InternalNameProvider nameProvider,
                                 final String generatedStorePrefix) {
+        this(materialized, nameProvider, generatedStorePrefix, null);
+    }
+
+    public MaterializedInternal(final Materialized<K, V, S> materialized,
+                                final InternalNameProvider nameProvider,
+                                final String generatedStorePrefix,
+                                final StreamsConfig config) {
         super(materialized);
 
+        if (config != null && this.storeSupplier == null && this.storeImplementation != null) {
+            // get default store implementation
+            this.storeImplementation = config.defaultStoreImplementation();
+        }
```

Comment:
new constructor for `MaterializedInternal`, to accept one more paremeter: `StreamConfig`, so that we can store the `storeImplementation` setting for later initialization.